tools/dep_tree_resolver/go_deps.go

```diff
@@ -0,0 +1,413 @@
+		moduleVersion, err := getActualModuleVersion(runCtx, moduleName)
+		if err != nil {
+			fmt.Printf("%v", err)
+			os.Exit(1)
+		}
+
+		if moduleVersion.Path == "" || moduleVersion.Version == "" {
+			fmt.Printf("ERROR! Module %s had an unresolved real version (%v)!", moduleVersion)
+			os.Exit(1)
+		}
+
+		modVersionMapping[moduleName] = moduleVersion
+
+		fmt.Printf("(%3d/%3d) %s -> %s\n", index+1, len(moduleNames), moduleName, moduleVersion)
+	}
+
+	return modVersionMapping, nil
+}
+
+func getDependencies(ctx context.Context) (string, []ModuleDep, map[string]Module, error) {
+	if Debug {
+		fmt.Println("Resolving the root package name...")
+	}
+
+	rootModule, err := getRootModule(ctx)
+	if err != nil {
+		return "", nil, nil, err
+	}
+
+	if Debug {
+		fmt.Printf("Root module resolved: %s\n", rootModule)
+		fmt.Println("Getting the dependency graph...")
+	}
+
+	depFlatGraph, moduleNames, err := runGraph(ctx, rootModule)
+	if err != nil {
+		return "", nil, nil, err
+	}
+
+	if Debug {
+		for _, module := range depFlatGraph {
+			fmt.Printf(
+				"%s@%s %s@%s\n",
+				module.Parent.Path,
+				module.Parent.Version,
+				module.Child.Path,
+				module.Child.Version,
+			)
+		}
+		fmt.Printf("Resolving the actual version used for modules (%d)...\n", len(moduleNames))
+	}
+
+	actualModuleVersions, err := resolveActualVersions(ctx, moduleNames)
+	if err != nil {
+		return "", nil, nil, err
+	}
+
+	if Debug {
+		fmt.Printf("Resolved %d modules\n", len(actualModuleVersions))
+	}
+
+	return rootModule, depFlatGraph, actualModuleVersions, nil
+}
+
+func getModuleDependencies(depFlatGraph []ModuleDep, module Module) []Module {
+	dependencies := []Module{}
+	for _, moduleDep := range depFlatGraph {
+		if moduleDep.Parent == module {
+			dependencies = append(dependencies, moduleDep.Child)
+		}
+	}
+	return dependencies
+}
+
+func isCirccularDependency(module *Module, depPath *[]Module) bool {
+	for _, seenModule := range *depPath {
+		if (*module).Path == seenModule.Path {
+			return true
+		}
+	}
+
+	return false
+}
+
+func resolveRecursive(
+	module *Module,
+	level int,
+	rootDepPath []Module,
+	depFlatGraph []ModuleDep,
+	actualModuleVersions map[string]Module,
+) (*DependencyTree, error) {
+
+	modPath := module.Path
+	if Debug {
+		fmt.Printf("Resolving dependency (Depth: %d) %s\n", level, modPath)
+	}
+
+	actualModule, ok := actualModuleVersions[modPath]
+	if !ok {
+		return nil,
+			fmt.Errorf("ERROR! Could not find actual module mapping for %s (level %d). Path: %+v", modPath, level, rootDepPath)
+	}
+
+	modDeps := getModuleDependencies(depFlatGraph, actualModule)
+
+	if Debug {
+		fmt.Printf("- Found %d dep(s) for module %s\n", len(modDeps), actualModule)
+	}
+
+	dependencies := []DependencyTree{}
+	for _, depModule := range modDeps {
+		// Circular dependencies are ignored once we encounter a loop
+		if isCirccularDependency(&depModule, &rootDepPath) {
+			if Debug {
+				fmt.Printf("WARN: Circular dependency breakout for %s\n", depModule.Path)
```

Comment:
The warning is just for debug level. It's not normally visible and is pretty valuable in figuring out if there's something wrong with the logic since the whole tree under the exclusion gets omitted.